{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7000d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "## information resources\n",
    "\n",
    "# CDS database fasta generated by NCBI Nucleotide\n",
    "cds=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/PaePhageInfoRetrieval/PaePhages_CDSs.fasta\"\n",
    "# ICTV VMR metadata export for Pseudomonas phages\n",
    "meta=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/PaePhageInfoRetrieval/ICTV_Metadata_raw.txt\"\n",
    "# Virus-host DB\n",
    "hosts=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/PaePhageInfoRetrieval/virushostdb.tsv\"\n",
    "\n",
    "## result files\n",
    "nhmmer=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/Hmmer1/GNAT/ICTV_nhmmer_GNAT.out.table\"\n",
    "phmmer=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/Hmmer2/phmmer.out\"\n",
    "supfam=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/SUPERFAMILY_interpro/Results_IDs\"\n",
    "hhblits=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/HHblits/counts.n3\"\n",
    "phmmer_revisit=\"/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/RevisitWithTP/phmmer/phmmer.TP.out\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19f52894",
   "metadata": {},
   "source": [
    "## Loading databases & defining auxiliary gathering function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5ba387f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2417c9cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_db = SeqIO.to_dict(SeqIO.parse(cds, \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "334f5040",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(meta, \"r\") as meta_handle:\n",
    "    meta_db = meta_handle.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d137e2a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(hosts, \"r\") as hosts_handle:\n",
    "    hosts_db = hosts_handle.read().split(\"\\n\")[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a684faac",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Collects the protein sequences of all CDS accessions in res_db in a new fasta file and composes a metadata report\n",
    "# @INPUTS\n",
    "# res_db: dictionary with the fasta name as key, and SeqRecord objects as values;\n",
    "#         similarly as the output of the to_dict() function of SeqIO\n",
    "# new_fasta_name: name of the new fasta file with all collected sequences\n",
    "# protein_list_name: name of the metadata report text file\n",
    "def gatherSequences(res_db, new_fasta_name, protein_list_name):\n",
    "    with open(protein_list_name, \"w\") as list_handle:\n",
    "        # header\n",
    "        list_handle.write(\"# Species\\tSpecies accession\\tProtein ID\\tAnnotation\\tHost\\tLength\\tLocation\\tStrand\\n\")\n",
    "        records=[]\n",
    "        for d in res_db.keys():\n",
    "            # The description field of the CDS fasta contains many metadata that are ready to scrape\n",
    "            description = '|'.join(res_db[d].description.split('|')[1:])\n",
    "            # protein annotation\n",
    "            protein =!echo \"$description\" | grep -Eo \"\\[protein=[^]]+?\\]\" | tr -d \"[]\" | cut -c 9-\n",
    "            protein = protein[0]\n",
    "            # genomic location\n",
    "            location =!echo \"$description\" | grep -Eo \"[0-9]+\\.\\.[0-9]+)?\\]\" | tr -d \"])\" | tr -s '.' '\\t'\n",
    "            location = [int(i) for i in location[0].split('\\t')]\n",
    "            # size in AA\n",
    "            length = str(int((location[1] - location[0]) / 3))\n",
    "            location = str(location)\n",
    "            # which DNA strand\n",
    "            strand =!echo \"$description\" | grep \"complement\"\n",
    "            if len(strand) == 1:\n",
    "                strand = \"-\"\n",
    "            else:\n",
    "                strand = \"+\"\n",
    "            # Protein accession\n",
    "            protein_id =!echo \"$description\" | grep -Eo \"protein_id=[A-Z|0-9]+\\.[0-9]\" | cut -c 12-\n",
    "            protein_id = protein_id[0]\n",
    "            # Species accession\n",
    "            species_accession =!echo \"$description\" | grep -Eo \"[A-Z|0-9]+?\\.[0-9]_prot\" | cut -d '.' -f 1\n",
    "            species_accession = species_accession[0]\n",
    "            # Common phage name\n",
    "            species =!echo -e \"$meta_db\" | grep \"$species_accession\" | cut -f 10\n",
    "            species = species[0]\n",
    "            # Official phage name\n",
    "            classification_name = !echo -e \"$meta_db\" | grep \"$species_accession\" | cut -f 9\n",
    "            classification_name = classification_name[0]\n",
    "            # Phage host\n",
    "            host = [i for i in hosts_db if species in i or classification_name in i]\n",
    "            if len(host) == 0:\n",
    "                species = species.replace(\"phage\", \"virus\")\n",
    "                host = [i for i in hosts_db if species in i or classification_name in i]\n",
    "                if len(host) == 0:\n",
    "                    print(species + \" not in hosts database\")\n",
    "                else:\n",
    "                    host = host[0]\n",
    "                    host =!echo -e \"$host\" | cut -f 9\n",
    "                    host = host[0]\n",
    "            else:\n",
    "                host = host[0]\n",
    "                host =!echo -e \"$host\" | cut -f 9\n",
    "                host = host[0]\n",
    "            \n",
    "            # write line in metadata report text file\n",
    "            list_handle.write(species + \"\\t\" + species_accession + \"\\t\" + protein_id + \"\\t\" + protein + \"\\t\" \\\n",
    "                              + host + \"\\t\" + length + \"\\t\" + location + \"\\t\" + strand + \"\\n\")\n",
    "            \n",
    "            # create fasta record in memory\n",
    "            new_id = species.split(' ')[2] + \"_\" + str(location).replace(', ','-').replace('[','').replace(']','') \\\n",
    "                        + \" (\" + species_accession + \")\"\n",
    "            record = SeqIO.SeqRecord(res_db[d].seq, id = new_id, name = new_id, description = protein_id)\n",
    "            records.append(record)\n",
    "    \n",
    "    # Sort metadata entries by phage name\n",
    "    !grep '#' $protein_list_name > header\n",
    "    !grep -v '#' $protein_list_name | sort > tmp\n",
    "    !rm $protein_list_name\n",
    "    !cat header tmp > $protein_list_name\n",
    "    !rm tmp\n",
    "    !rm header\n",
    "\n",
    "    # Write fasta file from records in memory\n",
    "    with open(new_fasta_name, \"w\") as handle:\n",
    "        SeqIO.write(records, handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "420abde5",
   "metadata": {},
   "source": [
    "### SUPERFAMILY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa2662a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract CDS keys from result file\n",
    "with open(supfam, \"r\") as supfam_handle:\n",
    "    supfam_keys = supfam_handle.read().split(\"\\n\")[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c456f82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the SeqRecord instances from the full database using the CDS keys\n",
    "supfam_seq_db = {i: seq_db[i] for i in supfam_keys}\n",
    "supfam_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c41cba5b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate multi-fasta and metadata report text file\n",
    "gatherSequences(supfam_seq_db, \"supfam.fasta\", \"protein_list_supfam\")\n",
    "!cat protein_list_supfam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "613a008d",
   "metadata": {},
   "source": [
    "### phmmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "661fb4c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(phmmer, \"r\") as phmmer_handle:\n",
    "    phmmer_keys_raw = phmmer_handle.read()\n",
    "    phmmer_keys =!echo -e \"$phmmer_keys_raw\" | grep -v \"#\" | cut -f 1\n",
    "    phmmer_keys = phmmer_keys[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f582ccd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "phmmer_seq_db = {i : seq_db[i] for i in phmmer_keys}\n",
    "phmmer_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76d0a323",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gatherSequences(phmmer_seq_db, \"phmmer.fasta\", \"protein_list_phmmer\")\n",
    "!cat protein_list_phmmer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9509f05",
   "metadata": {},
   "source": [
    " ### HHblits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686229d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(hhblits, \"r\") as hhblits_handle:\n",
    "    hhblits_keys_raw = hhblits_handle.read()\n",
    "    hhblits_keys =!echo -e \"$hhblits_keys_raw\" | grep -v '#' | cut -f 1 | cut -d '.' -f 1-3\n",
    "    hhblits_keys = hhblits_keys[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f66fe878",
   "metadata": {},
   "outputs": [],
   "source": [
    "hhblits_seq_db = {i: seq_db[i] for i in hhblits_keys}\n",
    "hhblits_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c54e13e0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gatherSequences(hhblits_seq_db, \"hhblits.fasta\", \"protein_list_hhblits\")\n",
    "!cat protein_list_hhblits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96aa3f29",
   "metadata": {},
   "source": [
    "### nhmmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d63c953",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Species accessions for all Pseudomonas phages\n",
    "all_species_acc = '/mnt/DATA/School/2022-2023/Thesis/Scripting/Mining/PaePhageInfoRetrieval/ICTV_Metadata_PaePhageIDs.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecc53ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### nhmmer hits are assigned to the CDS with which they have the most overlap in genomic location.\n",
    "\n",
    "## getting genomic locations of all CDSs\n",
    "# Read species accessions\n",
    "with open(all_species_acc, \"r\") as spec_acc_handle:\n",
    "    species = spec_acc_handle.read().split(\"\\n\")[:-1]\n",
    "with open(cds, \"r\") as cds_db_handle:\n",
    "    # Read full CDS fasta\n",
    "    cds_db_full = cds_db_handle.read().split(\"\\n\")[:-1]\n",
    "    # Extract headers\n",
    "    cds_db = [i for i in cds_db_full if '>' in i]\n",
    "    # Forget full CDS fasta to clear up memory\n",
    "    del cds_db_full\n",
    "    \n",
    "    # Create dictionary of lists of genomic locations by phage species accession\n",
    "    intervals = {}\n",
    "    for s in species:\n",
    "        # Get all CDS header strings for this species\n",
    "        cds_db_spec = [i for i in cds_db if s in i]\n",
    "        if len(cds_db_spec) == 0:\n",
    "            continue\n",
    "        # Convert list of CDS header strings into multi-line string\n",
    "        cds_db_str = '\\n'.join(cds_db_spec)\n",
    "        # Extract genomic locations and create list of genomic intervals for this species\n",
    "        location_raw =!echo \"$cds_db_str\" | grep -Eo \"[0-9]+\\.\\.[0-9]+)?\\]\" | tr -d \"])\" | tr -s '.' '\\t'\n",
    "        location = [[int(i) for i in j.split(\"\\t\")] for j in location_raw]\n",
    "        # add dictionary entry for this species\n",
    "        intervals[s] = location\n",
    "\n",
    "## getting genomic locations of nhmmer hits\n",
    "with open(nhmmer, \"r\") as nhmmer_handle:\n",
    "    nhmmer_raw = nhmmer_handle.read()\n",
    "    # Extract nhmmer hit species accessions\n",
    "    nhmmer_spec =!echo -e \"$nhmmer_raw\" | grep -v '#' | tr -s ' ' '\\t' | cut -f 1\n",
    "    nhmmer_spec = nhmmer_spec[:-1]\n",
    "    # Get start and end point of each nhmmer hit\n",
    "    nhmmer_int_start =!echo -e \"$nhmmer_raw\" | grep -v '#' | tr -s ' ' '\\t' | cut -f 7\n",
    "    nhmmer_int_start = [int(i) for i in nhmmer_int_start[:-1]]\n",
    "    nhmmer_int_end =!echo -e \"$nhmmer_raw\" | grep -v '#' | tr -s ' ' '\\t' | cut -f 8\n",
    "    nhmmer_int_end = [int(i) for i in nhmmer_int_end[:-1]]\n",
    "    # Convert into list of genomic intervals, turn around intervals of the complementary strand if required\n",
    "    nhmmer_int_list = [sorted([nhmmer_int_start[i], nhmmer_int_end[i]]) for i in range(len(nhmmer_int_start))]\n",
    "    \n",
    "## match genomic intervals by greatest overlap\n",
    "nhmmer_keys=[]\n",
    "for i in range(len(nhmmer_spec)):\n",
    "    # find genomic intervals of the species of this nhmmer hit\n",
    "    try:\n",
    "        cds_locs = intervals[nhmmer_spec[i]]\n",
    "    except KeyError:\n",
    "        print(\"No CDS found for \" + nhmmer_spec[i])\n",
    "        continue\n",
    "    # determine overlap for all CDSs of that species\n",
    "    cds_overlap = [min(nhmmer_int_list[i][1], cds_locs[j][1]) - max(nhmmer_int_list[i][0], cds_locs[j][0]) \\\n",
    "                   for j in range(len(cds_locs))]\n",
    "    # convert genomic location of CDS with largest overlap to NCBI Nucleotide-style genomic location string\n",
    "    cds_loc = '..'.join([str(i) for i in cds_locs[cds_overlap.index(max(cds_overlap))]])\n",
    "    # link to its CDS key via its header string\n",
    "    cds_key_header = [i for i in cds_db if cds_loc in i][0]\n",
    "    cds_key =!echo \"$cds_key_header\" | cut -d ' ' -f 1 | tr -d '>'\n",
    "    # add to the list of assigned CDS keys\n",
    "    nhmmer_keys.append(cds_key[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5240fdcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "nhmmer_seq_db = {i: seq_db[i] for i in nhmmer_keys}\n",
    "nhmmer_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04073942",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gatherSequences(nhmmer_seq_db, \"nhmmer.fasta\", \"protein_list_nhmmer\")\n",
    "!cat protein_list_nhmmer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44a234b9",
   "metadata": {},
   "source": [
    "### All together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe3e184",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_keys = list(set(supfam_keys + phmmer_keys + hhblits_keys + nhmmer_keys))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f414b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_seq_db = {i: seq_db[i] for i in all_keys}\n",
    "all_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77a0c851",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gatherSequences(all_seq_db, \"all.fasta\", \"protein_list_all\")\n",
    "!cat protein_list_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c401ee90",
   "metadata": {},
   "outputs": [],
   "source": [
    "!grep -v '#' protein_list_all | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df14ee8",
   "metadata": {},
   "source": [
    "Substracting the 31 known AceT, this make 81 newly found hits!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfe8e401",
   "metadata": {},
   "source": [
    "### Find new hits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f115ca45",
   "metadata": {},
   "source": [
    "First remove the initial set of AceT already known from Gene by Protein IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df5dd776",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cat ./Hmmer1/nhmmer.in.prot_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55599d6a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_all | tail -n +2 | sort > tmp1\n",
    "!sort ./Hmmer1/nhmmer.in.prot_IDs > tmp2\n",
    "!comm -23 tmp1 tmp2 > all.new\n",
    "!rm tmp1\n",
    "!rm tmp2\n",
    "!cat all.new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7e709cc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_nhmmer | tail -n +2 | sort > tmp1\n",
    "!sort ./Hmmer1/nhmmer.in.prot_IDs > tmp2\n",
    "!comm -23 tmp1 tmp2 > nhmmer.new\n",
    "!rm tmp1\n",
    "!rm tmp2\n",
    "!cat nhmmer.new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f9dd2c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_phmmer | tail -n +2 | sort > tmp1\n",
    "!sort ./Hmmer1/nhmmer.in.prot_IDs > tmp2\n",
    "!comm -23 tmp1 tmp2 > phmmer.new\n",
    "!rm tmp1\n",
    "!rm tmp2\n",
    "!cat phmmer.new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "714481c1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_supfam | tail -n +2 | sort > tmp1\n",
    "!sort ./Hmmer1/nhmmer.in.prot_IDs > tmp2\n",
    "!comm -23 tmp1 tmp2 > supfam.new\n",
    "!rm tmp1\n",
    "!rm tmp2\n",
    "!cat supfam.new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70842e10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_hhblits | tail -n +2 | sort > tmp1\n",
    "!sort ./Hmmer1/nhmmer.in.prot_IDs > tmp2\n",
    "!comm -23 tmp1 tmp2 > hhblits.new\n",
    "!rm tmp1\n",
    "!rm tmp2\n",
    "!cat hhblits.new"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d204351f",
   "metadata": {},
   "source": [
    "### phmmer revisit using TPs only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b59bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(phmmer_revisit, \"r\") as phmmer_revisit_handle:\n",
    "    phmmer_revisit_keys_raw = phmmer_revisit_handle.read()\n",
    "    phmmer_revisit_keys =!echo -e \"$phmmer_revisit_keys_raw\" | grep -v \"#\" | cut -f 1\n",
    "    phmmer_revisit_keys = phmmer_revisit_keys[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5444924",
   "metadata": {},
   "outputs": [],
   "source": [
    "phmmer_revisit_seq_db = {i: seq_db[i] for i in phmmer_revisit_keys}\n",
    "phmmer_revisit_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f15dd6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gatherSequences(phmmer_revisit_seq_db, \"phmmer.revisit.fasta\", \"protein_list_phmmer_revisit\")\n",
    "!cat protein_list_phmmer_revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "309c5542",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!comm -23 protein_list_all protein_list_phmmer_revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "530ebb18",
   "metadata": {},
   "outputs": [],
   "source": [
    "!comm -13 protein_list_all protein_list_phmmer_revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdbfcd66",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wc -l protein_list_phmmer_revisit\n",
    "!wc -l protein_list_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b3d219",
   "metadata": {},
   "source": [
    "### Preparing a TP-only fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4086bf8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_all | tail -n +2 | sort > tmp1\n",
    "!sort false.positives | comm -23 tmp1 - > all.TP\n",
    "!rm tmp1\n",
    "!cat all.TP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0cc8a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wc -l all.TP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e7050c",
   "metadata": {},
   "outputs": [],
   "source": [
    "TP_accession =!cat all.TP\n",
    "cds_accession = [i.split(\"_\")[2] for i in list(seq_db.keys())]\n",
    "TP_indices = [i for i in range(len(cds_accession)) if cds_accession[i] in TP_accession]\n",
    "all_TP_keys = [list(seq_db.keys())[i] for i in TP_indices]\n",
    "all_TP_seq_db = {i: seq_db[i] for i in all_TP_keys}\n",
    "all_TP_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11a59e8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gatherSequences(all_TP_seq_db, \"all.TP.fasta\", \"protein_list_all_TP\")\n",
    "!cat protein_list_all_TP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1038da72",
   "metadata": {},
   "source": [
    "### Preparing a TP-only fasta after the phmmer revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77fe0f48",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!cut -f 3 protein_list_all | tail -n +2 | sort > tmp1\n",
    "!comm -13 protein_list_all protein_list_phmmer_revisit | cut -f 3 > tmp2\n",
    "!sort false.positives | comm -23 tmp1 - > tmp3\n",
    "!cat tmp3 tmp2 > all.TP.revisit\n",
    "!rm tmp1\n",
    "!rm tmp2\n",
    "!rm tmp3\n",
    "!cat all.TP.revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2a14aad",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wc -l all.TP.revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3ac0fd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wc -l false.positives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f156e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "TP_revisit_accession =!cat all.TP.revisit\n",
    "TP_revisit_indices = [i for i in range(len(cds_accession)) if cds_accession[i] in TP_revisit_accession]\n",
    "all_TP_revisit_keys = [list(seq_db.keys())[i] for i in TP_revisit_indices]\n",
    "all_TP_revisit_seq_db = {i: seq_db[i] for i in all_TP_revisit_keys}\n",
    "all_TP_revisit_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "798ce6f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gatherSequences(all_TP_revisit_seq_db, \"all.TP.revisit.fasta\", \"protein_list_all_TP_revisit\")\n",
    "!cat protein_list_all_TP_revisit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c83d39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
